Support service update in BPF maps

tgraf opened the pull request:
The existing servive update logic deleted and re-added a service every time a
service had to be updated. This commit adds update logic to the load balancer
map and uses it when adding and updating services.

Related: #5425

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5469)
<!-- Reviewable:end -->

aanm commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -21,13 +21,20 @@ import (
+
+	// mutex protects access to the BPF map to guarantee atomicity if a
+	// transaction must be split across multiple map access operations.
+	mutex lock.RWMutex
```
this can be a `lock.Mutex` as the only RLock operations happen during a delete.

tgraf replied:
Both dump operations (`DumpServiceMapsToUserspace()`, `DumpRevNATMapsToUserspace()`) take a read lock so they can run in parallel which is definitely valuable. Do you disagree?

aanm replied:
the `lock.Mutex` vs `lock.RWMutex` speed gain is just a matter of how many "reads" one expects the map will have. How frequently will `DumpServiceMapsToUserspace` and `DumpRevNATMapsToUserspace` that justify a `RWMutex` over a `Mutex`?

tgraf replied:
I think it's more than that and what you describe mostly applies if the lock is taking very often.

We need to separate between:

1. Lock is taken very often for short periods of time and mostly for writing and only sometimes for reading. In this case, using Mutex makes sense.
2. Lock is taken rarely but for longer period of times. In this case, it makes sense to use RWMutex even if the ratio of reads vs writes is 1:10 as blocking out all readers for a long time is hurting a lot.

The dump operations are by far the most expensive operation as we will perform a system call for nservice*nbackend. By using an RWLock, we can dump in parallel. Right now, we dump in the k8s LB sync controller and in the blocking SyncLB(). Being able to run the most expensive operations in parallel is a clear benefit already. The benefit is currently limited by the overarching loadbalancer lock which serializes everything so the benefit is currently limited but that though. We should definitely revise the loadbalancer lock though and only hold it to protect the in-memory map.

aanm replied:
oh ok, that makes more sense then

aanm commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -313,27 +325,55 @@ func UpdateWrrSeq(fe ServiceKey, weights []uint16) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	svcValue, err := lookupService(fe)
```
Do we still want to continue if err is not nil? I guess so, if the service does not exist. Can you add a comment here if that's the case?

tgraf replied:
Added comment

aanm reviewed (dismissed):
Small nits. The overall logic seems correct

tgraf committed:
lbmap: Support transactional updates

The existing servive update logic deleted and re-added a service every time a
service had to be updated. This commit adds update logic to the load balancer
map and uses it when adding and updating services.

There is still a small race window:

1. If the number of service backends is reduced
2. AND the CPU processing a packet in the datapath gets preempted between
   looking up the master key and looking up the selected slave.
3. AND the selected slave has an index that is greater than the new number of
   backends
4. THEN the slave lookup will fail

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lbmap: Introduce lock to allow for transactional operations

There is a high level lock in daemon/loadbalanger.go which lbmap currently
relies on but this is very fragile and not ideal. Introduce a lock on lbmap
level. It's a simple global lock spanning all maps for now. It could eventually
become more fine grained to cover only certain maps at a time but the
contention on this lock is minimal for now.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lbmap: Mark internal APIs as private

Signed-off-by: Thomas Graf <thomas@cilium.io>
